### PR TITLE
SEAB-5104: Add Notebook entity and related stuff

### PR DESCRIPTION
**Pull request by svonworl**
**Description**
This PR adds a `Notebook` entity and and an associated DAO to the webservice, creates the new `notebook` db table, adds a new descriptor language and some subclass and file type values, and modifies related code and entities so that the webservice still builds.

Please keep in mind that this is a first attempt at something that will provide a suitable foundation for the rest of the work.  Given the unknowns, it's probably somewhere in the neighborhood of reasonable, but it definitely embodies some educated guesses.   We can iterate if we discover that we need to make changes/improvements...

My current plan is to try to shoehorn the "format" of the notebook (IPYNB, RStudio, etc) and the programming language used within (Python, R, Julia, etc) into the notions of an entry's descriptor language and descriptor language subclass, respectively.  This may make it easier to reuse various bits of code, as the expense of not being quite as understandable as storing some or all of that info in custom field(s) of the concrete entities.

To that end, we:
* define a new descriptor language `IPYNB`.
* add values for Python, R, and Julia to `DescriptorLanguageSubclass`.
* add new file types `DOCKSTORE_IPYNB` (notebook file), `DOCKSTORE_REES` (REES configuration files), and `DOCKSTORE_DATA` (data and other accompanying files from the repo that the notebook uses).
* add new file type category `CONFIGURATION_FILE` to describe ancillary configuration files such as those specified by REES.
 
We won't support R and Julia initially, but I put them in there to add meaning/context.  I can take them out if we find it objectionable...

The enum value names are likely to change as we do more work, these are a first draft.

I didn't include the notion of Docker kernel image in the definition of a notebook [version] in this PR, due to the feedback from the PI meeting and the fact since it's currently unclear how high we are prioritizing the "on-prem" use case.  We can add it later if we need to.
 
Most of the new code is [intentionally] a pretty straightforward adaptation of existing `AppTool` and `Service` code.

A very rudimentary test, adapted from `ServiceIT`, which slightly exercises the notebook DAO, is included.  We will add more tests as we build the surrounding functionality... 

**Review Instructions**
Check `openapi.yaml` to make sure the new descriptor language values have appeared, that there is a new `notebook` table in the database, and that from the point of view of the end user, nothing has changed.

**Issue**
https://ucsc-cgl.atlassian.net/browse/SEAB-5104

**Security**
If there are any concerns that require extra attention from the security team, highlight them here.

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future
- [x] If this PR is for a user-facing feature, create and link a documentation ticket for this feature (usually in the same milestone as the linked issue). Style points if you create a documentation PR directly and link that instead.

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Notebook.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+package io.dockstore.webservice.core;
+
+import io.dockstore.common.EntryType;
+import io.swagger.annotations.ApiModel;
+import javax.persistence.Entity;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+
+@ApiModel(value = "Notebook", description = "This describes one notebook in the dockstore as a special generate case of a workflow", parent = Workflow.class)
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Notebook.java:25]: "generate" -> "degenerate"? If copied from Service

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Workflow.java [hunk 2]
```diff
@@ -131,13 +131,13 @@ public abstract class Workflow extends Entry<Workflow, WorkflowVersion> {
     // this one is annoying since the codegen doesn't seem to pick up @JsonValue in the DescriptorLanguage enum
     @Column(nullable = false)
     @Convert(converter = DescriptorLanguageConverter.class)
-    @ApiModelProperty(value = "This is a descriptor type for the workflow, by default either SMK, CWL, WDL, NFL, or gxformat2 (Defaults to CWL).", required = true, position = 18, allowableValues = "SMK, CWL, WDL, NFL, gxformat2, service")
+    @ApiModelProperty(value = "This is a descriptor type for the workflow, by default either SMK, CWL, WDL, NFL, or gxformat2 (Defaults to CWL).", required = true, position = 18, allowableValues = "SMK, CWL, WDL, NFL, gxformat2, service, ipynb")
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Workflow.java:134]: Value was not updated -- perhaps intentionally to not publicize at this point?

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/NotebookIT.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
+/*
+ *    Copyright 2019 OICR
```
Comment (coverbeck) [inline dockstore-integration-testing/src/test/java/io/dockstore/webservice/NotebookIT.java:2]: Copyright year

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSNotebookPath.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2022 OICR, UCSC
```
Comment (coverbeck) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/database/RSSNotebookPath.java:2]: Off by one in the year

File: dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java [hunk 5]
```diff
@@ -244,7 +251,8 @@ public enum FileType {
         DOCKSTORE_YML(FileTypeCategory.OTHER),
         DOCKSTORE_SERVICE_YML(FileTypeCategory.PRIMARY_DESCRIPTOR), DOCKSTORE_SERVICE_TEST_JSON(FileTypeCategory.TEST_FILE), DOCKSTORE_SERVICE_OTHER(FileTypeCategory.OTHER),
         DOCKSTORE_GXFORMAT2(FileTypeCategory.GENERIC_DESCRIPTOR), GXFORMAT2_TEST_FILE(FileTypeCategory.TEST_FILE),
-        DOCKSTORE_SWL(FileTypeCategory.GENERIC_DESCRIPTOR), SWL_TEST_JSON(FileTypeCategory.TEST_FILE);
+        DOCKSTORE_SWL(FileTypeCategory.GENERIC_DESCRIPTOR), SWL_TEST_JSON(FileTypeCategory.TEST_FILE),
+        DOCKSTORE_IPYNB(FileTypeCategory.PRIMARY_DESCRIPTOR), DOCKSTORE_REES(FileTypeCategory.CONFIGURATION_FILE), DOCKSTORE_DATA(FileTypeCategory.OTHER);
```
Comment (coverbeck) [inline dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java:255]: Nit, DOCKSTORE_DATA seems a little too generic to me; I'd scope it down to something more notebook specific, if it makes sense. IOW, I can tell what the other file types are used for by their name only, this one is a little vaguer. DOCKSTORE_NOTEBOOK_DATA? May not be precise, and maybe there's not a better alternative.

**Review by coverbeck — APPROVED** [review]

**Comment by denis-yuen** [comment]
> include the notion of Docker kernel image in the definition of a notebook [version] in this PR ... since it's currently unclear how high we are prioritizing the "on-prem" use case.

The result is fine, but it's unclear to me how these two ideas are related. Can be Slack message if confidential

File: dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java [hunk 6]
```diff
@@ -80,6 +80,12 @@ public boolean isRelevantFileType(FileType type) {
         public boolean isRelevantFileType(FileType type) {
             return super.isRelevantFileType(type) || type == FileType.DOCKSTORE_SERVICE_OTHER;
         }
+    },
+    IPYNB("ipynb", "Jupyter IPYNB notebook", FileType.DOCKSTORE_IPYNB, null, false, false, Set.of("ipynb"), false, false) {
```
Comment (denis-yuen) [inline dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java:84]: second last parameter implies that this won't support hosted, intentional?

**Review by denis-yuen — COMMENTED** [review]
on build success since we're changing the DB, admittedly we seem to be having issues, but I may have some ideas .... https://ucsc-cgl.atlassian.net/browse/SEAB-5163

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Workflow.java (hunk 2 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/Workflow.java:134]: > Value was not updated -- perhaps intentionally to not publicize at this point?

yup

File: dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java (hunk 6 above)
Comment (svonworl) [reply dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java:84]: Yeah, if eventually, we either provide a local notebooks environment, or we figure out a way to launch our hosted content to a notebook platform, we'll probably toggle this to `true`.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java [hunk 7]
```diff
@@ -118,6 +124,7 @@ public class Event {
 
     public Event() { }
 
+    /*
```
Comment (kathy-t) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java:127]: Hmm, big comment block. Should this constructor be deleted?

**Comment by svonworl** [comment]
> > include the notion of Docker kernel image in the definition of a notebook [version] in this PR ... since it's currently unclear how high we are prioritizing the "on-prem" use case.
> 
> The result is fine, but it's unclear to me how these two ideas are related. Can be Slack message if confidential

[Clarified in Slack]

**Review by kathy-t — APPROVED** [review]

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java (hunk 7 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java:127]: Oh, yeah, I think so, this was part of an experiment that I forgot to reverse...

**Review by denis-yuen — APPROVED** [review]

File: dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java (hunk 5 above)
Comment (svonworl) [reply dockstore-common/src/main/java/io/dockstore/common/DescriptorLanguage.java]: Changed to `DOCKSTORE_NOTEBOOK_REES` and `DOCKSTORE_NOTEBOOK_OTHER`, respectively, the latter of which is consistent with the naming scheme in the `Service` implementation...

File: dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java (hunk 7 above)
Comment (svonworl) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/core/Event.java]: Removed the [commented-out] big constructor.

**Review by david4096 — APPROVED** [review]

File: dockstore-webservice/src/main/resources/swagger.yaml [hunk 8]
```diff
@@ -9042,12 +9058,15 @@ definitions:
...
-          \ it is only used for services."
+          \ it is only used for services and notebooks."
         enum:
         - "DOCKER_COMPOSE"
         - "HELM"
         - "SWARM"
         - "KUBERNETES"
+        - "PYTHON"
+        - "R"
+        - "JULIA"
```
Comment (david4096) [inline dockstore-webservice/src/main/resources/swagger.yaml:9069]: Nice, but I'm guessing aspirational

Comment (svonworl) [reply dockstore-webservice/src/main/resources/swagger.yaml:9069]: yuppers

**Pull request merged**